active/0002-centalized-mgmt-resource-types-provider-versions.md

```diff
@@ -0,0 +1,247 @@
+
+```bicep
+import 'sys@1.0.0'
+import 'kubernetes@1.0.0' as k8s
+```
+
+#### Change the delimiter '@' used by the supported provider declaration syntax to ':' to align with module reference syntax and increase consistency
+
+The syntax shown below uses ':' instead of '@' as a delimiter for provider declaration statements
+
+```bicep
+import 'br:mcr.microsoft.com/bicep/providers/az:0.2.3' // notice its NOT using az@0.2.3
+import 'br/public:az:0.2.3' as myAz
+```
+
+## Drawbacks
+
+> Discuss any potential drawbacks associated with this approach. Why might this _not_ be a suitable choice? Consider implementation costs, integration with existing and planned features, and the migration cost for Bicep users. Clearly identify if it constitutes a breaking change.
+
+1. The proposal uses `bicepconfig.json` to declare the provider information and is ruled by `bicepconfig.json` merge semantics. The merge semantics for `bicepconfig.json` specify that for a given Bicep source file, the configuration file closest to the Bicep file in the directory hierarchy is used. This behavior is not intuitive and may cause users that have specified multiple config files in their project file-structure to experience unexpected behaviors that may include build errors when Bicep attempts to resolve a namespace identifier. This is illustrated by the example below.
+
+**Example: Provider resolution error due as a result of `bicepconfig.json` merge semantics**
+
+Consider a Bicep project under folder `src` with the following files:
+
+- `src/main.bicep`
+
+```bicep
+provider foo as mainFooProvider
+```
+
+- `src/bicepconfig.json`
+
+```bicep
+{
+    "providers":{
+        "foo": {
+            "source": {
+                "registry": "mcr.microsoft.com/bicep/providers/az",
+                "version": "1.2.3",
+            },
+            "isDefaultImport": false
+        },
+        "bar" : {
+            "source": {
+                "registry": "mcr.microsoft.com",
+                "repository": "bicep/providers",
+                "name":
+                "version": "3.2.1"
+            },
+            "isDefaultImport": false
+        }
+    }
+}
+```
+
+- `src/modules/mod.bicep`
+
+```bicep
+provider bar as modBarProvider
+provider foo as modFooProvider
+```
+
+- `src/modules/bicepconfig.json`
+
+```bicep
+{
+     "experimentalFeaturesEnabled": {
+        "compileTimeImports": false
+     },
+}
+```
+
+The user may intuitively expect Bicep to resolve the provider details of `modFooProvider` by inspection of `src/bicepconfig.json` and for this project to compile successfully in a way that feels familiar with other programming languages where merge semantics are applied on code dependencies (e.g. NPMs `package.json`), however, Bicep config merge semantics uses the configuration file closest to the Bicep file in the directory hierarchy and merges them to the builtin config.
```

Comment:
W.r.t familiarity - is this true? I didn't know that NPM had merge semantics.

Reply:
I need to double check, I may be mistaken here for sue as I am not a JavaScript expert. In any case the example is to illustrate and idea. Perhaps I can find a better example for this

Reply:
The best prior art on merging that I recall from previous discussion is [the `"extends"` property in tsconfig.json](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html#tsconfig-bases). NPM doesn't perform any kind of merge, though there are some tools that let you do merging (like Lerna).

Reply:
rephrased to remove the example of package.json as I learned that npm doesn't support merge semantics natively